File: validateSAID.scala
object ValidateSAID {
 
  def check(value: String): Int = {
    val id = value.trim()
    if (!id.matches("[0-9]+")) {
      -1
    }
    if (id.length != 10) {
      -1
    }
    val idType: Int = id.substring(0, 1).toInt
    if (idType != 2 && idType != 1) {
      -1
    }
    var sum: Int = 0
    for (i <- 0.until(10)) {
      if (i % 2 == 0) {
        val ZFOdd: String = "%02d".format(
         (id.substring(i, i + 1)).toInt * 2)
        sum += (ZFOdd.substring(0, 1)).toInt + (ZFOdd.substring(1, 2)).toInt
      } else {
        sum += (id.substring(i, i + 1)).toInt
      }
    }
    if ((sum % 10 != 0)) -1 else idType
  }

}

File: validateSAID.dart
class ValidateSAID {
  static final RegExp _numericRegex = RegExp(r'^\d+$');
  static const int _idLength = 10;
  static const int _type1 = 1;
  static const int _type2 = 2;

  int check(String id) {
    id = id.trim();
	  int type = int.parse(id.substring(0, 1));
    if (!_numericRegex.hasMatch(id) || id.length != _idLength || (type != _type1 && type != _type2)) {
      return -1;
    }
	  int sum = 0;
    for (int i = 0; i < _idLength; i++) {	  
	  if (i.isEven) {
        String ZFOdd = "${int.parse(id.substring(i, i+1)) * 2}".padLeft(2, '0');
        sum += int.parse(ZFOdd.substring(0, 1)) + int.parse(ZFOdd.substring(1, 2));
      } else {
        sum += int.parse(id.substring(i, i+1));
      }
    }
    return (sum % 10 != 0) ? -1 : type;
  }
  
}
